lib/translator/inquiry.cc

```diff
@@ -0,0 +1,253 @@
+  char product_serial_number[100] = {0}, hex_string[33];
+  const size_t kNguidLen = 40, kEui64Len = 20, kV1SerialLen = 30;
+
+  // check if nonzero
+  bool nguid_nz =
+      identify_namespace_data.nguid[0] || identify_namespace_data.nguid[1];
+  bool eui64_nz = identify_namespace_data.eui64;
+
+  if (nguid_nz || eui64_nz) {
+    if (nguid_nz) {
+      // 6.1.3.1.1
+      result.page_length = kNguidLen;
+
+      // convert 128-bit number into hex string, 64-bits at a time
+      sprintf(hex_string, "%08lx", identify_namespace_data.nguid[0]);
+      sprintf(&hex_string[16], "%08lx", identify_namespace_data.nguid[1]);
+
+    } else if (eui64_nz) {
+      // 6.1.3.1.2
+      result.page_length = kEui64Len;
+
+      // convert 64-bit number into hex string
+      sprintf(hex_string, "%08lx", identify_namespace_data.eui64);
+    }
+
+    // insert _ and . in the correct positions
+    for (int i = 4; i < result.page_length; i += 5) {
+      product_serial_number[i] = '_';
+    }
+    product_serial_number[result.page_length - 1] = '.';
+
+    // insert numbers
+    int pos = 0;
+    for (int i = 0; i < result.page_length - 1; i++) {
+      if (product_serial_number[i] != '_') {
+        product_serial_number[i] = hex_string[pos++];
+      }
+    }
+  } else {
+    // 6.1.3.1.3
+    // valid for NVMe 1.0 devices only
+
+    // PAGE LENGTH should be set to 30 indicating the page length in bytes.
+    result.page_length = kV1SerialLen;
+
+    // PRODUCT SERIAL NUMBER should be formed as follows:
+    // Bits 239:80 20 bytes of Serial Number (bytes 23:04 of Identify
+    // Controller data structure)
+    static_assert(sizeof(identify_controller_data.sn) == 20);
+    memcpy(&product_serial_number, identify_controller_data.sn,
+           sizeof(identify_controller_data.sn));
+
+    // Bits 79:72 ASCII representation of “_"
+    product_serial_number[sizeof(identify_controller_data.sn)] = '_';
+
+    // Bits 71:08 ASCII representation of 32 bit Namespace Identifier (NSID)
+    sprintf(&product_serial_number[sizeof(identify_controller_data.sn) + 1],
+            "%08x", nsid);
+
+    // Bits 07:00 ASCII representation of “."
+    product_serial_number[kV1SerialLen - 1] = '.';
+  }
+
+  memcpy(buffer.data(), &result, sizeof(result));
+  memcpy(&buffer[sizeof(result)], &product_serial_number, result.page_length);
+}
+
+StatusCode InquiryToNvme(absl::Span<const uint8_t> scsi_cmd,
+                         nvme_defs::GenericQueueEntryCmd &identify_ns,
+                         nvme_defs::GenericQueueEntryCmd &identify_ctrl,
+                         uint32_t &alloc_len) {
+  scsi_defs::InquiryCommand cmd =
+      *reinterpret_cast<const scsi_defs::InquiryCommand *>(&scsi_cmd[1]);
+  alloc_len = cmd.allocation_length;
+  // Allocate prp & assign to command
+  void *prp_ptr = AllocPages(1);
+  if (prp_ptr == nullptr) return StatusCode::kFailure;
+  uint64_t prp = reinterpret_cast<uint64_t>(prp_ptr);
```

Comment:
nit: is it possible to name this `ns_prp` and the other one `ctrl_prp` instead of `prp` and `prp2` just so it's clearer that they'refor two different commands